### PR TITLE
feat: ensure `permanentId` field exists and correctly configured

### PR DESCRIPTION
**Pull request by y-lakhdar**
To address the issue raised in [discuss](https://discuss.coveo.com/t/log-click-events-default-contentidkey-values-permanentid-vs-permanentid-vs-urihash/7356/6).

## Main changes
1. Throws an error if the `permanentid` field has been created with a wrong type.
2. Creates the `permanentid` field if it does not exist in the org.

## Other changes
1 Convert the [fieldBuilder.ts](https://github.com/coveo/push-api-client.js/compare/CDX-851...CDX-850?expand=1#diff-ca96c846624702481a136353d22416cb1f2f8d9fd82e43378e45f6edacaadf08) into a Map. It's way more cleaner that way.

### CODE REVIEW

**Review by louis-bompart — APPROVED** [review]

File: src/fieldAnalyser/fieldBuilder.ts [hunk 1]
```diff
@@ -1,24 +1,19 @@
 import {FieldModel, FieldTypes} from '@coveord/platform-client';
 
-export class FieldBuilder {
-  public readonly dict: Record<string, FieldTypes> = {};
-
-  constructor() {}
-  public add(name: string, type: FieldTypes) {
-    this.dict[name] = type;
-    return this;
+export class FieldBuilder extends Map<string, FieldTypes> {
```
Comment (louis-bompart) [inline src/fieldAnalyser/fieldBuilder.ts:3]: Mh the change is not that inconsequential here. Is a FieldBuilder really a map?
In that case maybe it should have a name that explain that it's just a data holding structure.

File: src/errors/fieldErrors.ts [hunk 2]
```diff
@@ -22,3 +23,17 @@ export class FieldTypeInconsistencyError extends PushApiClientBaseError {
...
+export class InvalidPermanentId extends PushApiClientBaseError {
+  public name = 'Invalid permanentId field Error';
+  public constructor(field: FieldModel) {
+    super(`
+    The detected permanentid field in the index is incorrectly configured.
+    Expected field type: STRING
+    Current field type: ${field.type}
+
+    You can delete and recreate the permanentid field as a STRING field.
+    See https://docs.coveo.com/en/1913/leverage-machine-learning/about-the-permanentid-field fore more info.
```
Comment (jpmarceau) [inline src/errors/fieldErrors.ts:36]: (just for brevity:)
```suggestion
    See docs.coveo.com/en/1913 fore more info.
```

**Review by olamothe — APPROVED** [review]

File: src/errors/fieldErrors.ts [hunk 3]
```diff
@@ -22,3 +23,17 @@ export class FieldTypeInconsistencyError extends PushApiClientBaseError {
       Make sure to review and fix your document metadata before pushing`;
   }
 }
+
+export class InvalidPermanentId extends PushApiClientBaseError {
+  public name = 'Invalid permanentId field Error';
+  public constructor(field: FieldModel) {
+    super(`
+    The detected permanentid field in the index is incorrectly configured.
```
Comment (jpmarceau) [inline src/errors/fieldErrors.ts:31]: (just a bit easier to read:)
```suggestion
    The permanentid field detected in the index is not correctly configured.
```

**Review by jpmarceau — APPROVED** [review]
👍

File: src/fieldAnalyser/fieldBuilder.ts (hunk 1 above)
Comment (y-lakhdar) [reply src/fieldAnalyser/fieldBuilder.ts:3]: Yes, it's really just a map holding data. I agree, the name is misleading.

File: src/errors/fieldErrors.ts (hunk 3 above)
Comment (y-lakhdar) [reply src/errors/fieldErrors.ts]: Thanks!

**Pull request merged**